File: ui/src/logic/filters.js
export const filters = {
  genre: (album, val) => {
    if (Array.isArray(album.GENRE)) {
      return album.GENRE.includes(val);
    }
    return album.GENRE === val;
  },

  totaltracks: (album, val) => album.total_tracks === String(val),

  search: (album, val) => {
    const q = val.toLowerCase();

    if (
      (album.ALBUM && album.ALBUM.toLowerCase().includes(q)) ||
      (album.ALBUMARTIST && album.ALBUMARTIST.toLowerCase().includes(q))
    ) {
      return true;
    }

    if (album.tracks && album.tracks.some(t => t.TITLE && t.TITLE.toLowerCase().includes(q))) {
      return true;
    }

    return false;
  },

  decade: (album, val) => {
    if (!album.DATE) return false;
    const year = parseInt(album.DATE.substring(0, 4));
    const start = parseInt(val.substring(0, 4));
    const end = start + 9;
    return year >= start && year <= end;
  },

  recent: (album) => {
    const now = Math.floor(Date.now() / 1000);
    const added = parseInt(album.unix_added || 0);
    return (now - added) < (86400 * 30);
  },

  year_added: (album, val) => {
    const unix = parseInt(album.unix_added || 0);
    if (unix <= 0) return false;
    const year = new Date(unix * 1000).getFullYear().toString();
    return year === val;
  },

  month_added: (album, val) => {
    const unix = parseInt(album.unix_added || 0);
    if (unix <= 0) return false;
    const date = new Date(unix * 1000);
    const mVal = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;
    return mVal === val;
  },

  chroma: (album, val) => {
    const score = parseFloat(album.cover_chroma || 0);
    switch (val) {
      case "Monochrome": return score === 0;
      case "Bleak":      return score > 0 && score < 15;
      case "Muted":      return score >= 15 && score < 33;
      case "Standard":   return score >= 33 && score < 60;
      case "Vibrant":    return score >= 60;
      default:           return true;
    }
  }
};

export function applyFilter(album, filterKey, filterVal) {
  if (!filterKey || !filters[filterKey]) return true;
  return filters[filterKey](album, filterVal);
}

File: rust/src/egui/logic/filters.rs
use crate::server::library::models::AlbumView;

pub fn apply_filter(album: &AlbumView, key: &str, val: &str) -> bool {
    match key {
        "genre" => {
            album.album_data.genre.iter().any(|g| g.to_lowercase() == val.to_lowercase())
        }
        "decade" => {
            if album.album_data.date.len() < 4 { return false; }
            let year_str = &album.album_data.date[0..4];
            let start_str = &val[0..4];
            if let (Ok(year), Ok(start)) = (year_str.parse::<u32>(), start_str.parse::<u32>()) {
                year >= start && year <= start + 9
            } else {
                false
            }
        }
        "year_added" => {
            let year = chrono::DateTime::from_timestamp(album.album_data.info.unix_added as i64, 0)
                .map(|dt| dt.format("%Y").to_string())
                .unwrap_or_default();
            year == val
        }
        "chroma" => {
            let score = album.album_data.tags.get("COVER_CHROMA")
                .and_then(|v| v.as_f64())
                .unwrap_or(0.0);
            match val {
                "Monochrome" => score == 0.0,
                "Bleak" => score > 0.0 && score < 15.0,
                "Muted" => score >= 15.0 && score < 33.0,
                "Standard" => score >= 33.0 && score < 60.0,
                "Vibrant" => score >= 60.0,
                _ => true,
            }
        }
        "search" => {
            let q = val.to_lowercase();
            if album.album_data.album.to_lowercase().contains(&q) || 
               album.album_data.albumartist.to_lowercase().contains(&q) {
                return true;
            }
            album.tracks.iter().any(|t| t.title.to_lowercase().contains(&q))
        }
        _ => true,
    }
}
